### PR TITLE
fix: Write article about invisible unicode characters TG-453

### PR DESCRIPTION
**Pull request by stepan662**

### CODE REVIEW

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+---
+title: Using invisible text for better localization
+authors: [sgranat]
+tags: [SDK, i18n, i18next]
+---
+
+![Text with ghosts](/img/blog/invisible-text/invisible_text.png)
+
+Some time ago I've came across an [article](https://null-byte.wonderhowto.com/how-to/use-zero-width-characters-hide-secret-messages-text-even-reveal-leaks-0198692/) about inserting secret text into regular messages using invisible zero-width unicode characters. Now during my work on localization SDKs, I've found quite interesting usecase for it.
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:9]: use case

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+<!--truncate-->
+
+## Background
+
+At tolgee we are offering localization SDKs, which enable you insert your translated keys into web application. One of our main selling points is that we offer in-context localization out of box. In-context localization means, that when you are in development mode, you can click on any translation in the app and change the translation directly in the app.
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:15]: Tolgee

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+To make this work our SDKs needs to somehow mark where the translations are. We might just search the dom for occurences of translations in given language, however that won't cover more advanced cases, when we for example need to support variables in translations. Then the easiest way is to wrap each translation witch `span` element and give it some special attribute, so then we can find it. Easy peasy, no? Well there is a catch ... what if user needs to translate texts where you can't use HTML (e.g. element attributes)? Or what if the extra `span` element is not acceptable?
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:17]: DOM, with 'span'

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+> #### Why is context important?
+> When you translate the app, it's important that you see where exactly is the translation located as pure translation from one language to another can be incorrect in many cases. So we offer in-context translation directly in the App and we want this feature to work everywhere.
+
+### Text based solution
+
+Our original solution was in these cases insert string with our special syntax, instead of the translation self. Then we would observe whole page with [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) and when we encounter this "encoded" text we find parent element, and store the information about the translation and replace this message with our translation. As we found out this approach works quite well, mainly because observer is triggered before changes are rendered to the page, so this whole replacement is invisble to the user.
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:24]: translation itself?

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+However, we still run into problems sometimes. We usually want to run observer only on `body`, so if you update e.g. page title, so then you can see text which is completely broken. Also when you use it in cases, when you need to measure the width of an element right after you change it - this will cause you problems, because the element is containing different text, so the measurements will be different. We offer the user an option to turn wrapping off, but he needs to do that explicitly and then the in-context doesn't work. Not ideal.
+
+![Broken translation](/img/blog/invisible-text/encoded_wrong.png)
+
+## The idea of invisible marks
+
+I think you can already see how invisible marks might be useful to us. I'll get to how it works, but for now let's just imagine, we have ability to insert invisible information to any text and then retrieve it. We can now insert secrets into every translation. Same as with regular text we can still observe the dom for changes and when we find this "watermark" we know which translation it's containing and in-context can work anywhere.
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:32]: DOM

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+### Text based solution
+
+Our original solution was in these cases insert string with our special syntax, instead of the translation self. Then we would observe whole page with [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) and when we encounter this "encoded" text we find parent element, and store the information about the translation and replace this message with our translation. As we found out this approach works quite well, mainly because observer is triggered before changes are rendered to the page, so this whole replacement is invisble to the user.
+
+However, we still run into problems sometimes. We usually want to run observer only on `body`, so if you update e.g. page title, so then you can see text which is completely broken. Also when you use it in cases, when you need to measure the width of an element right after you change it - this will cause you problems, because the element is containing different text, so the measurements will be different. We offer the user an option to turn wrapping off, but he needs to do that explicitly and then the in-context doesn't work. Not ideal.
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:26]: in-context localization?

File: blog/2021-12-17-invisible-characters-for-better-localization.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+One caveat is that `ZWJ` causes two characters acting as one, so it might have effect on text surrounding it. In our usecase when we use multiple `ZWN` and `ZWJ` are used in succession, only the last one is interpreted so my solution is to use one extra `ZWN` at the end of each byte, so it's always last and that should eliminate all side effects.
+
+The fact that each secret byte is 9 characters long is making long texts extremely memory inefficient. So I've choosed to assign each translation a number and then encode it as unicode character. This way I take an advantage of utf-8 characters which can be 1-4 bytes long, so I have effective space about 1 milion combinations, which is more than enough for one page and I'll use relatively small amount of invisible characters (in most cases 9 or 18 for one key).
```
Comment (marketachalupnikova) [inline blog/2021-12-17-invisible-characters-for-better-localization.md:46]: million

File: blog/2021-12-17-invisible-characters-for-better-localization.md (hunk 2 above)
Comment (JanCizmar) [reply blog/2021-12-17-invisible-characters-for-better-localization.md:15]: "which enable you +to"?

in the app too many times?
on any translation -`in the app` and change the translation directly in the app.

File: blog/2021-12-17-invisible-characters-for-better-localization.md (hunk 3 above)
Comment (JanCizmar) [reply blog/2021-12-17-invisible-characters-for-better-localization.md:17]: To make this work+, our SDKs ne

File: blog/2021-12-17-invisible-characters-for-better-localization.md (hunk 4 above)
Comment (JanCizmar) [reply blog/2021-12-17-invisible-characters-for-better-localization.md:24]: -, instead

**Review by JanCizmar — APPROVED** [review]

**Pull request merged**